sympy/tensor/array/sp_utils.py

```diff
@@ -0,0 +1,345 @@
+    """
+    def __new__(cls, iterable=None, shape=None, **kwargs):
+        shape, flat_list = cls._handle_ndarray_creation_inputs(iterable, shape, **kwargs)
+        #shape = Tuple(*map(_sympify, shape))
+        shape = Tuple.fromiter(_sympify(i) for i in shape)
+        cls._check_special_bounds(flat_list, shape)
+        loop_size = functools.reduce(lambda x,y: x*y, shape) if shape else 0
+
+        # Ideally we should use the Array module for this empty initialization, but
+        # it is not yet supported in SymPy
+        data = cls._empty(shape)
+        rows = cls._empty(shape)
+
+        self = Basic.__new__(cls, data, rows, shape, **kwargs)
+        self._shape = shape
+        self._rank = len(shape)
+        self._loop_size = loop_size
+
+        new_dict = {}
+        # if it is a dense array, it would be cast to a default sparse array
+        # and then convert to a LIL format
+        if isinstance(iterable, Iterable):
+            iterable = SparseNDimArray(iterable)
+
+        if isinstance(iterable, SparseNDimArray):
+            new_dict = iterable._sparse_array
+
+        # TODO: Enable the initialization with (data, rows) as a tuple
+        else:
+            raise NotImplementedError("Data type not yet supported")
+
+        # Add non zero value to internal list.
+        # This operation can be simplified once Array module supports tuple index
+        for k, v in new_dict.items():
+            idx = self._get_tuple_index(k)
+            self._get_row(data, idx).append(v)
+            self._get_row(rows, idx).append(idx[-1])
+
+        self._data = data
+        self._rows = rows
+        return self
+
+    @classmethod
+    def _empty(cls, shape):
+        def f(s):
+            if len(s) == 0:
+                return []
+            if len(s) == 1:
+                return [[] for i in range(s[0])]
+            arr = f(s[1:])
+            return [arr for i in range(s[0])]
+
+        if not shape:
+            raise ValueError("Shape must be defined")
+        return f(shape[:-1])
+
+    def _get_row(self, iterable, index):
+        temp_iter = iterable
+        for i in index[:-1]:
+            temp_iter = temp_iter[i]
+        return temp_iter
+
+    def __iter__(self):
+        def iterator():
+            for i in range(self._loop_size):
+                yield self[i]
+        return iterator()
+
+    def __getitem__(self, index):
+        if not isinstance(index, (tuple, Tuple)):
+            index = self._get_tuple_index(index)
+        if not check_bound(self._shape, index):
+            raise ValueError('Index ' + str(index) + ' out of border')
+
+        row_values = self._get_row(self._data, index)
+        row_indices = self._get_row(self._rows, index)
+
+        if index[-1] in row_indices:
+            value_index = row_indices.index(index[-1])
+            return row_values[value_index]
+        else:
+            return 0
+
+    # For mutable arrays
+    def __setitem__(self, index, value):
```

Comment:
This should raise an exception if the class inherits `Basic`

Reply:
Yes, it is a little bit contradictory to handle both immutable and mutable property in one class. I was thinking that we can handle this property by adding a flag: `isMutable`, but it turns out that it is not very pratical.

Reply:
No, it cannot be done. The reason is complicated but SymPy needs all of its objects to be immutable. If you want a mutable object, you need to create a separate class.

Have a look at the existing class structures.